RulesEngine/Ruleset.cs

```diff
@@ -81,217 +67,306 @@ public void AddDirectory(string path, string tag = null)
                 }
-                throw new Exception($"Error parsing {sourcename}: {ex.Message}");
+
+                _oatRules.Add(cor);
             }
+        }
 
-            if (ruleList != null)
-            {
-                foreach (Rule r in ruleList)
-                {
-                    if (_uniqueRuleIds.Add(r.Id))
-                    {
-                        r.Source = sourcename;
-                        r.RuntimeTag = tag;
+        /// <summary>
+        ///     Load rules from JSON string
+        /// </summary>
+        /// <param name="jsonstring"> JSON string </param>
+        /// <param name="sourcename"> Name of the source (file, stream, etc..) </param>
+        /// <param name="tag"> Tag for the rules </param>
+        public void AddString(string jsonstring, string sourcename, string? tag = null)
+        {
+            AddRange(StringToRules(jsonstring, sourcename, tag));
+        }
 
-                        if (r.Patterns == null)
-                        {
-                            r.Patterns = new SearchPattern[] { };
-                        }
+        /// <summary>
+        ///     Filters rules within Ruleset by languages
+        /// </summary>
+        /// <param name="languages"> Languages </param>
+        /// <returns> Filtered rules </returns>
+        public IEnumerable<ConvertedOatRule> ByLanguages(string[] languages)
+        {
+            return _oatRules.Where(x => x.AppInspectorRule.AppliesTo is null || x.AppInspectorRule.AppliesTo.Length == 0 || (x.AppInspectorRule.AppliesTo is string[] appliesList && appliesList.Any(y => languages.Contains(y))));
+        }
 
-                        foreach (SearchPattern pattern in r.Patterns)
-                        {
-                            SanitizePatternRegex(pattern);
-                        }
+        /// <summary>
+        ///     Filters rules within Ruleset by languages
+        /// </summary>
+        /// <param name="languages"> Languages </param>
+        /// <returns> Filtered rules </returns>
+        public IEnumerable<ConvertedOatRule> ByLanguage(string language)
+        {
+            return _oatRules.Where(x => x.AppInspectorRule.AppliesTo is null || x.AppInspectorRule.AppliesTo.Length == 0 || (x.AppInspectorRule.AppliesTo is string[] appliesList && appliesList.Any(y => language.Contains(y))));
+        }
 
-                        if (r.Conditions == null)
-                        {
-                            r.Conditions = new SearchCondition[] { };
-                        }
+   
+        public ConvertedOatRule? AppInspectorRuleToOatRule(Rule rule)
+        {
+            if (rule == null)
+                return null;
 
-                        foreach (SearchCondition condition in r.Conditions)
-                        {
-                            SanitizePatternRegex(condition.Pattern);
-                        }
+            _rules.Add(rule); //add native Rule to avoid creating from enumerated _oatRules later
```

Comment:
This is a pretty nasty side effect of calling this function.  I'd put this elsewhere if you want to keep a stash of the original rules.

Reply:
The counts always matched but maybe your thinking of the use of the yield in StringToRules so I moved it in a minimalist way to resolve.  Thanks.

Reply:
The rules happened to match because you happened to only call this function once per rule. The function says "convertthis" it shouldn't have other side effects other than returning the converted object. Unless the function is "convert this and add it to a queue first" but then you need to change the name of the function. 

Imagine you want to test at some point and just want to call this function to convert. You never intended to change the state of the rules field you just wanted the converted object to test with.

Reply:
Unless you access _rules extensively and absolutely need the performance (I don't think this is the case) my recommendation would be to ditch keeping track of them this way at all and use Linq to .Select the rules out when you want them. Just change _rules to be a property with 
```csharp
get{
    return _oatRules.Select(x => x.AppInspectorRule);
}
```